tests/default.nix

```diff
@@ -0,0 +1,32 @@
+{ pkgs, inputs }:
+pkgs.testers.runNixOSTest {
+  name = "boot";
+  node.specialArgs = {
+    inherit inputs;
+  };
+  nodes.machine = { lib, config, ... }: {
+    imports = [
+      ../configuration.nix
+
+      # The `virtualisation.vmVariant` setting we to import VM-specific settings
+      # doesn't for the test VMs.
+      # https://github.com/NixOS/nixpkgs/pull/339511#issuecomment-2328611982
+      ../vm.nix
+    ];
+
+    services.packit-api.instances =
+      let
+        f = name: lib.nameValuePair name {
+          authentication.method = lib.mkForce "basic";
+        };
+      in
+      lib.listToAttrs (map f config.services.multi-packit.instances);
+
+    # It's suprisingly easy to run qemu without hardware acceleration and not
+    # notice it, which makes the VM so slow the tests tend to fail. This forces
+    # KVM acceleration and will fail to start if missing.
+    virtualisation.qemu.options = [ "-machine" "accel=kvm" ];
```

Comment:
okay been reading up about kvm and qemu, i get the general gist but couldnt find exactly how kvm acceleration works, is it because qemu simulates a full machine and kvm hardware acceleration brings that abstraction closer to the metal of the actual machine the vm is running on?

Reply:
I'm not an expert, but here's my attempt: QEMU is a general virtual machine framework. It has multiple backends, mostly TCG and KVM.

[TCG](https://wiki.qemu.org/Documentation/TCG) is an actual emulator, implemented as a JIT. It reads instructions from the guest machine and translates them into host machine instructions. It also needs to emulate a whole bunch of low level stuff (eg. the memory managment, interrupts, ...). Emulating stuff this way is super slow. Fine if you want to emulate a 90s video game console on a modern fast CPU, less fine if you want to emulate a modern fast CPU on a modern fast CPU.

KVM is the linux API for hardware accelerated VM. It's the Linux equivalent of Hyper-V. It uses whatever the underlying CPU's acceleration is, on intel that's Intel VT. In that mode, guest instructions are run directly on the host CPU. The CPU has special settings to run in VM mode, and these instructions are correctly isolated from the host and each other. The performance of this is close to native (there's usually a small overhead, but not much). It's what everyone.

Because KVM needs special hardware access, some linux distros, including Ubuntu, restrict the permissions a little bit. Not all though, from my reading it seems some of them have `/dev/kvm` as 666, so writable by everyone.

I spent way too long figuring out why the tests were slow but `nix .#start-vm` was fast. Turns out it was because the former runs inside the nix sandbox and doesn't have my permissions. This line makes it so that the tests fail with an error that is obvious, instead of failing because they time out for being way too slow.